src/Upgrade420.php

```diff
@@ -0,0 +1,307 @@
+<?php
+/**
+ * Upgrade 4.2.0
+ *
+ * @author    Pronamic <info@pronamic.eu>
+ * @copyright 2005-2022 Pronamic
+ * @license   GPL-3.0-or-later
+ * @package   Pronamic\WordPress\Pay\Upgrades
+ */
+
+namespace Pronamic\WordPress\Pay\Extensions\WooCommerce;
+
+use Pronamic\WordPress\Pay\Upgrades\Upgrade;
+
+/**
+ * Upgrade 4.2.0
+ *
+ * @author  Reüel van der Steege
+ * @version 4.2.0
+ * @since   4.2.0
+ */
+class Upgrade420 extends Upgrade {
+	/**
+	 * Construct 4.2.0 upgrade.
+	 */
+	public function __construct() {
+		parent::__construct( '4.2.0' );
+
+		if ( \defined( '\WP_CLI' ) && \WP_CLI ) {
+			$this->cli_init();
+		}
+	}
+
+	/**
+	 * Execute.
+	 *
+	 * @return void
+	 */
+	public function execute() {
+		$this->upgrade_subscriptions();
+	}
+
+	/**
+	 * WP-CLI initialize.
+	 *
+	 * @link https://github.com/wp-cli/wp-cli/issues/4818
+	 * @return void
+	 */
+	public function cli_init() {
+		\WP_CLI::add_command(
+			'pronamic-pay woocommerce upgrade-420 execute',
+			function ( $args, $assoc_args ) {
+				\WP_CLI::log( 'Upgrade 4.2.0' );
+
+				$this->execute();
+			},
+			[
+				'shortdesc' => 'Execute WooCommerce upgrade 4.2.0.',
+			]
+		);
+
+		\WP_CLI::add_command(
+			'pronamic-pay woocommerce upgrade-420 list-subscriptions',
+			function ( $args, $assoc_args ) {
+				\WP_CLI::log( 'Upgrade 4.2.0 - Subscriptions List' );
+
+				$posts = $this->get_subscription_posts();
+
+				\WP_CLI\Utils\format_items( 'table', $posts, [ 'ID', 'post_title', 'post_status' ] );
+			},
+			[
+				'shortdesc' => 'List subscriptions for WooCommerce upgrade 4.2.0.',
+			]
+		);
+
+		\WP_CLI::add_command(
+			'pronamic-pay woocommerce upgrade-420 upgrade-subscriptions',
+			function ( $args, $assoc_args ) {
+				\WP_CLI::log( 'Upgrade 4.2.0 - Subscriptions' );
+
+				$this->upgrade_subscriptions(
+					[
+						'dry-run'  => \WP_CLI\Utils\get_flag_value( $assoc_args, 'dry-run', true ),
+						'post__in' => \WP_CLI\Utils\get_flag_value( $assoc_args, 'post__in', null ),
+					]
+				);
+			},
+			[
+				'shortdesc' => 'Upgrade subscriptions for WooCommerce upgrade 4.2.0.',
+			]
+		);
+	}
+
+	/**
+	 * Log.
+	 *
+	 * @link https://make.wordpress.org/cli/handbook/internal-api/wp-cli-log/
+	 * @param string $message Message.
+	 * @return void
+	 */
+	private function log( string $message ) {
+		if ( method_exists( '\WP_CLI', 'log' ) ) {
+			\WP_CLI::log( $message );
+		}
+	}
+
+	/**
+	 * Get subscription posts to upgrade.
+	 *
+	 * @param array $args Query arguments.
+	 * @return array
+	 */
+	private function get_subscription_posts( $args = [] ) {
+		$args['post_type']     = 'pronamic_pay_subscr';
+		$args['post_status']   = 'any';
+		$args['nopaging']      = true;
```

Comment:
@remcotolsma would this be an issue for larger sites? Thoughts on how to improve?

Reply:
@rvdsteege ideally we should set it up like in https://github.com/pronamic/wp-pay-core/blob/046ed470ff0966419fda8ae663a9630b95b0c70a/src/Subscriptions/SubscriptionsFollowUpPaymentsController.php#L141-L178 ?